### PR TITLE
Symfony 4 compatibility

### PR DESCRIPTION
**Pull request by martin-georgiev**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -13,17 +13,17 @@
...
-        "symfony/dependency-injection": "~2.0|~3.0",
-        "symfony/form": "~2.0|~3.0",
-        "symfony/http-foundation": "~2.0|~3.0",
-        "symfony/http-kernel": "~2.0|~3.0",
+        "symfony/browser-kit": "~2.0|~3.0|~4.0",
+        "symfony/config": "~2.0|~3.0|~4.0",
+        "symfony/dependency-injection": "~2.0|~3.0|~4.0",
+        "symfony/form": "~2.0|~3.0|~4.0",
+        "symfony/http-foundation": "~2.0|~3.0|~4.0",
+        "symfony/http-kernel": "~2.0|~3.0|~4.0",
```
Comment (weaverryan) [inline composer.json]: I think we can safely bump to `~2.7|~3.0|~4.0` to only support, still-supported Symfony versions

Comment (martin-georgiev) [reply composer.json]: Sounds reasonable, PHP min version now also moved up to 5.3.9 (enforced by Symfony 2.7)

**Comment by martin-georgiev** [comment]
@weaverryan Nice spot, the change in default service visibility was poorly neglected on my end.
I didn't make all services public as I believe there is no such use-case for `payment.paypal.authentication_strategy.token` and `payment.paypal.authentication_strategy`. Any other thoughts about them?

**Comment by weaverryan** [comment]
That’s up to the maintainer :). The easiest thing is to set them all to public=true, as it means there is no behavior change or BC break. So for the purpose of making this an “easy merge”, that’s the best path :)

**Comment by weaverryan** [comment]
@martin-georgiev Actually, I was wrong - @stof just corrected me :). If some services should be private in the future, then do not change them at all (i.e. do not give them a `public` attribute). We only need to add `public="true"` for services that we DO want to keep public.

For services that we would rather be private, if we do *not* add a `public` property, then it will become private in Symfony 4. But, this is *not* a BC break, because the user will get a deprecation notice if they try to use this service.  So in fact, I think your original implementation was actually ok :)

**Comment by martin-georgiev** [comment]
On a side node the package utilises PHP 7.2+ deprecations. A separate PR will be needed addressing that issues.
@weaverryan I'm confused what is the root cause for the [failure](https://travis-ci.org/schmittjoh/JMSPaymentPaypalBundle/builds/304037842) in Travis's PHP 7.1 + Symfony 4 matrix step. It's a second day I'm trying to replicate it locally with little to no success. Any ideas?

**Comment by regularjack** [comment]
Thank you for this, highly appreciated.

Looking at the passing builds ([example](https://travis-ci.org/schmittjoh/JMSPaymentPaypalBundle/jobs/304037846)), I'm puzzled by the fact that composer is able to resolve dependencies since [JMSPaymentCoreBundle](https://github.com/schmittjoh/JMSPaymentCoreBundle) requires `~2.3|~3.0`. Any idea why it's working? It definitely should not.

Independently of why the build is passing, before we add Symfony 4.0 support here, [JMSPaymentCoreBundle](https://github.com/schmittjoh/JMSPaymentCoreBundle) needs to support it. Is this something you would be available to do? I just merged https://github.com/schmittjoh/JMSPaymentCoreBundle/pull/220 which will make the job easier.

Finally, I would prefer if this PR would focus on only adding support for Symfony 4.0, without bumping minimum required versions of PHP and Symfony. As it is now, it's not clear for me which changes are required for Symfony 4.0 and which are optional. That said, I'm definitely willing to discuss bumping the minimum required versions but I would prefer that discussion to happen in another PR, it you're ok with it.

**Comment by martin-georgiev** [comment]
@regularjack In my understanding Composer dependencies are safely resolved because of the circular require reference in both packages (this is no longer the case after your merge of schmittjoh/JMSPaymentCoreBundle#220).

I've dropped the proposed bump of PHP 5.3.9 & Symfony 2.7 minimum versions as you are right that it blurs the changelog. I'll address these changes in a follow-up MR.

It's on my to-do list for next week to prepare JMSPaymentCoreBundle to be Symfony 4 ready so expect a MR for it too :).

**Comment by regularjack** [comment]
Closing in favour of #90

**Pull request closed without merging**